### PR TITLE
Prevent accidental deletion of copy when it's not been republished yet

### PR DESCRIPTION
**Pull request by enricobattocchi**
## Context

* Visiting `duplicatePost.originalEditURL` triggers the deletion of the R&R copy even if it's not been rewritten yet. We should avoid that.

## Summary

This PR can be summarized in the following changelog entry:

* Prevents accidentale deletion of a R&R copy via redirect if it's not been republished yet

## Relevant technical choices:

*

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

### Block editor:
* create a R&R copy and edit it on block editor
* using the console, copy the value for `duplicatePost.originalEditURL` and paste it as address for a new browser tab, then visit it
* see that you get a `wp_die` "An error occurred while deleting the Rewrite & Republish copy."
* visit the post list, or reload the tab where you were editing the copy, to see the copy is still there
* republish the copy via Republish button to see everything's fine (the original is overwritten and the copy deleted)

### Classic editor
Basically the Classic editor after rewriting the original triggers a `wp_safe_redirect` to the above URL so nothing is really changing, but you can test this way:
* comment out `$this->republish( $post, $original_post );` in class-post-republisher.php line 144
* activate Classic editor plugin
* create and edit a R&R copy
* republish it via button: since the original won't be rewritten because of the code commetend out, you will be redirected but the check will fail, so you'll get a `wp_die` "An error occurred while deleting the Rewrite & Republish copy."
* visit the post list to see the copy is still there
* remove the comment and republish the copy via Republish button to see everything's fine (the original is overwritten and the copy deleted)

### Test instructions for QA when the code is in the RC

* [ ] QA should use the same steps as above.

QA can test this PR by following these steps:

* N/A, this is to prevent some unwanted consequences that can arise with custom code or other plugins. Just testing for regression (you should be able to republish a copy both in Block editor and in Classic editor) is fine.

## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

*

## UI changes

* [ ] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Documentation

* [ ] I have written documentation for this change.

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [ ] I have added unittests to verify the code works as intended

Fixes [P3-322]

### CODE REVIEW

File: src/class-post-republisher.php [hunk 1]
```diff
@@ -216,7 +216,11 @@ public function clean_up_after_redirect() {
 
 			\check_admin_referer( 'dp-republish', 'dpnonce' );
 
-			$this->delete_copy( $copy_id, $post_id );
+			if ( \intval( \get_post_meta( $copy_id, '_dp_has_been_republished', true ) ) === 1 ) {
+				$this->delete_copy( $copy_id, $post_id );
+			} else {
```
Comment (afercia) [inline src/class-post-republisher.php:221]: A Yoast best practice is to avoid if/else when possible e,g, by returning within the if. Since Duplicate Post doesn't fully adopt the Yoast CS I guess this can stay for now.

**Comment by afercia** [comment]
CR 👍

**Comment by JesserH** [comment]
Acceptance 🚧 
When on classic editor and having the line commented out, the following happens:
I create a R&R copy and hit republish.
I get the following popup:
Screenshot 2021-01-18 at 11 23 19
Now when i visit the posts page, i notice that the copy does not exist no more:
Screenshot 2021-01-18 at 11 24 07

**Comment by afercia** [comment]
Acceptance 2 🚧 

It took me a while to understand what's going on because it's partly related to the edge case to follow in the test instructions but it's also something that needs to be double checked. To reproduce, follow these steps:

### Block editor:
* take note of the ID of the post you're going to create a R&R copy from
* create a R&R copy and edit it with the block editor
* using the console, copy the value for `duplicatePost.originalEditURL` and paste it in the address bar of a new browser tab, then visit it
* see that you get a `wp_die` "An error occurred while deleting the Rewrite & Republish copy."
* inspect the database > `wp_postmeta` table and search for the `post_id` of the original post
* you will get all the post meta related to the original post: observe there's no `_dp_has_been_republished` meta
* visit the post list, or reload the tab where you were editing the copy, to see the copy is still there
* republish the copy via the Republish button (the original is overwritten and the copy deleted)
* inspect the database again and see that this time the original post has a post meta _dp_has_been_republished with a value `1`

To my understanding, the `_dp_has_been_republished` should be set on the R&R copy. Not sure why it is set on the original post instead. I'd think this should never happen and we should make sure the _dp_has_been_republished post meta is set only on the copy.

At this point, if you follow the test instructions for Classic Editor **using the same original post**, when you create he R&R copy also the _dp_has_been_republished post meta will be copied into the copy. Thus, the copy will be already flagged as "being republished" and the check will fail. Even if you comment out the line in the test instructions, the copy will be "repubilshed" (actually the changes are not copied).

**Comment by afercia** [comment]
CR 2 & Acceptance 2 👍
